### PR TITLE
Implement the base class for the shared state.

### PR DESCRIPTION
**Pull request by coryan**
Futures and promises keep a shared state that tracks whether the promise
is satisfied, if the value has been retrieved, the exception (if any),
and hold the (type erased) continuation. In this PR we introduce this
base class and unit tests for it, as part of the work for #1345.

The class is incomplete, and there are a lot of support classes around
it that will be introduced in future PRs, as it is, this PR is getting
too large already.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1351)

### CODE REVIEW

**Comment by coryan** [comment]
To get a sense of where this is going this might be useful:

https://github.com/coryan/google-cloud-cpp/blob/9cb2bf2a02734aa30af31f352484caaedc1bf425/google/cloud/internal/future_void_then_test.cc#L29

But it will take at least 3 more PRs to get there I think.

**Comment by coryan** [comment]
/cc: @houglum @frankyn because you said you would be interested in taking a look.

File: google/cloud/internal/future_impl.h [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+   *
+   * @return `std::future_status::ready` if the shared state is satisfied.
+   *     `std::future_status::deferred` if the shared state is not satisfied and
+   *     there is a continuation ready to execute when it is satisfied.
+   *     `std::future_status::timeout` otherwise.
+   */
+  template <typename Rep, typename Period>
+  std::future_status wait_for(std::chrono::duration<Rep, Period> duration) {
+    std::unique_lock<std::mutex> lk(mu_);
+    if (not lk.owns_lock()) {
```
Comment (dopiera) [inline google/cloud/internal/future_impl.h]: Why this `if` necessary? If I'm reading the docs and code correctly, `lk` will acquire the mutex. Same question for `wait_until`.

File: google/cloud/internal/future_impl.h [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  void set_exception(std::exception_ptr ex, std::unique_lock<std::mutex>& lk) {
+    if (is_ready_unlocked()) {
+      throw std::future_error(std::future_errc::promise_already_satisfied);
+    }
+    exception_ = std::move(ex);
+    current_state_ = state::has_exception;
+  }
+
+  /// Immediately notify all waiting threads.
+  void notify_now(std::unique_lock<std::mutex>& lk) {
```
Comment (dopiera) [inline google/cloud/internal/future_impl.h]: I'm not a big fan of releasing a `mutex` in a different scope than it was acquired. IMO it makes it harder to read and given that this function is always called at the end of functions, you could skip the `unlock` here and change the signature to take the lock by a const reference.

Up to you, though.

File: google/cloud/internal/future_impl.h [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+  mutable std::mutex mu_;
+  std::condition_variable cv_;
+  enum class state {
+    not_ready,
+    has_exception,
+    has_value,
+  };
+  state current_state_;
+  std::exception_ptr exception_;
```
Comment (dopiera) [inline google/cloud/internal/future_impl.h]: If we used C++17 I'd have suggested a `std::variant` for `state_` and `exception_`, but it's probably not worth the time to hand-craft it.

File: google/cloud/internal/future_impl_test.cc [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+TEST(FutureImplBaseTest, Basic) {
+  future_shared_state_base shared_state;
+  EXPECT_FALSE(shared_state.is_ready());
+}
+
+TEST(FutureImplBaseTest, WaitFor) {
+  future_shared_state_base shared_state;
+  auto s = shared_state.wait_for(100_us);
+  EXPECT_EQ(static_cast<int>(s), static_cast<int>(std::future_status::timeout));
+  EXPECT_FALSE(shared_state.is_ready());
```
Comment (dopiera) [inline google/cloud/internal/future_impl_test.cc]: Maybe it's excessive, but maybe you could add a check that we actually spent at least 100us?

**Pull request merged**